### ISSUE TITLE
[🚀 Feature]: Create separate NPM for Selenium Manager

### ISSUE THREAD

**Issue by titusfortner** [SeleniumHQ/selenium#11378]
### Feature and motivation

Unlike the other bindings, there are various JavaScript projects (like webdriver.io) that can benefit from Selenium Manager being released as a separate package. @harsha509 mentioned it shouldn't be too challenging to figure out how to build it with Bazel. @christian-bromann might be able to provide feedback on requirements for this.

### Usage example

n/a

**Comment by christian-bromann** [comment]
Ideally it would be great if Selenium Manager could package a WASM file and publish it to NPM with a minimal JS interface to run the features of the manager, e.g. install driver etc. I recently was working on [a project](https://github.com/stateful/runmejs) with similar requirements. Happy to provide support if needed.

**Comment by titusfortner** [comment]
Would that require putting the Rust code in the package?

**Comment by christian-bromann** [comment]
> Would that require putting the Rust code in the package?

No. I recommend have a [`prepare`](https://github.com/stateful/runmejs/blob/main/package.json#L26) script in the `package.json` that automatically downloads the latest Selenium Manager release with the WASM file included. You can go all fancy and pin NPM release versions to the Selenium Manager release version and have auto-releases. Anyway the code for such NPM package can live totally independent of the original source.

**Comment by diemol** [comment]
This assumes that we are going to make Selenium Manager releases, which I would like to. But so far we do not have consensus of that. Are we actually going to make releases or not?

**Comment by titusfortner** [comment]
If supporting wdio is the primary argument in favor of independent release right now, I'd prefer finding an alternate way to support/distribute, at least until we have full functionality. 

Also, this isn't just one binary, but 3 because of different architectures, not sure how that works with npm.

**Comment by diemol** [comment]
We have two use cases now, WebdriverIO and Docker images setup.

**Comment by christian-bromann** [comment]
> Also, this isn't just one binary, but 3 because of different architectures, not sure how that works with npm.

It does, example case: https://github.com/webdriverio-community/node-geckodriver - the package downloads the binary from Github when the user uses the driver for the first time. Something similar could be done for an NPM package for Selenium Manager.

**Comment by titusfortner** [comment]
Docker case is more an optimization(?) than new functionality.

1. We need an API we can agree to maintain backwards compatible if we release independently. What we have now isn't what we want, and I'm not sure we'll know for sure what we want until browser downloads are implemented.
2. User confusion. I feel like we're just starting to make progress here :joy:
3. It seems like there are alternatives to separate downloads that are sufficient for now?

**Comment by diemol** [comment]
The Docker case is just to use the functionality that already exists in Selenium Manager

**Comment by titusfortner** [comment]
As I understand it, the Docker use case is to be able to load it as part of the Docker RUN command instead of the first time a session is requested by the entrypoint. There's no missing functionality right now with this scenario; Selenium Manager will just execute the first session request, it's just an optimization to start it at the beginning.

**Comment by diemol** [comment]
Ok, yes. Optimization.

**Comment by titusfortner** [comment]
But along those lines. Maybe we can document how people could execute selenium manager code to cache drivers without starting drivers? Not a use case I'd considered.

**Comment by titusfortner** [comment]
That last comment would be much more easily done when we implement — https://github.com/SeleniumHQ/selenium/issues/13022

**Comment by titusfortner** [comment]
> I recommend have a [prepare](https://github.com/stateful/runmejs/blob/main/package.json#L26) script in the package.json

@christian-bromann this is for a new "Selenium Manager" npm? If we release as part of Selenium, we can build / grab it dynamically without needing to have official releases. Can you provide an example file for what we need so we can figure out what it would take to release multiple npm packages with Bazel. (Simon is moving to updated rules which should make that easier)

**Comment by christian-bromann** [comment]
I recommend a simple wrapper module that downloads the binary if used the first time. See [this example](https://github.com/webdriverio-community/node-geckodriver/blob/main/src/install.ts#L34-L37) of the Geckodriver package. You can also include the binary to the NPM package but it feels like an unnecessary extra payload. I personally prefer NPM packages to be as small as possible.

**Comment by diemol** [comment]
During the May 2024 Selenium Summit, we made the decision to implement this.

**Comment by titusfortner** [comment]
Reopening: this was auto-closed by the stale bot, but it's actively being worked on. #17542 ("[py][js] Initial Creation of a external shippable selenium manager") implements the shippable packages, exposing Selenium Manager via `npx`/`uvx` with per-platform packages. It's part of #17326. Keeping this open to track the npm side.
